libvast/test/module.cpp

```diff
@@ -0,0 +1,689 @@
+        if (attribute_record.size() != 1)
+          return caf::make_error(ec::parse_error, "parses an attribute not "
+                                                  "having one and only one "
+                                                  "field");
+        const auto& attribute_key = attribute_record.begin()->first;
+        const auto* attribute_value_ptr
+          = caf::get_if<std::string>(&attribute_record.begin()->second);
+        if (attribute_value_ptr == nullptr)
+          return caf::make_error(ec::parse_error, "parses an attribute with a "
+                                                  "non-string value");
+        const auto& attribute_value = *attribute_value_ptr;
+        attributes.push_back({attribute_key, attribute_value});
+      }
+    }
+  }
+  // Check that only one of type, enum, list, map and record is specified
+  // by the user
+  auto found_type = decl.find("type");
+  auto found_enum = decl.find("enum");
+  auto found_list = decl.find("list");
+  auto found_map = decl.find("map");
+  auto found_record = decl.find("record");
+  auto is_type_found = found_type != decl.end();
+  auto is_enum_found = found_enum != decl.end();
+  auto is_list_found = found_list != decl.end();
+  auto is_map_found = found_map != decl.end();
+  auto is_record_found = found_record != decl.end();
+  int type_selector_cnt = 0;
+  if (is_type_found)
+    type_selector_cnt++;
+  if (is_enum_found)
+    type_selector_cnt++;
+  if (is_list_found)
+    type_selector_cnt++;
+  if (is_map_found)
+    type_selector_cnt++;
+  if (is_record_found)
+    type_selector_cnt++;
+  if (type_selector_cnt != 1)
+    return caf::make_error(ec::parse_error, "expects one of type, enum, list, "
+                                            "map, record");
+  // Type alias
+  if (is_type_found) {
+    auto type_expected = to_type(known_types, found_type->second);
+    if (!type_expected)
+      return type_expected.error();
+    return type{name, *type_expected, std::move(attributes)};
+  }
+  // Enumeration
+  if (is_enum_found) {
+    const auto* enum_list_ptr = caf::get_if<list>(&found_enum->second);
+    if (enum_list_ptr == nullptr)
+      return caf::make_error(ec::parse_error, "parses an enum with an invalid "
+                                              "format");
+    const auto& enum_list = *enum_list_ptr;
+    if (enum_list.empty())
+      return caf::make_error(ec::parse_error, "parses an empty enum");
+    auto enum_fields = std::vector<enumeration_type::field_view>{};
+    for (const auto& enum_value : enum_list) {
+      const auto* enum_string_ptr = caf::get_if<std::string>(&enum_value);
+      if (enum_string_ptr == nullptr)
+        return caf::make_error(ec::parse_error, "parses an enum value with an "
+                                                "invalid format");
+      enum_fields.push_back({*enum_string_ptr});
+    }
+    return type{name, enumeration_type{enum_fields}, std::move(attributes)};
+  }
```

Comment:
Can you extract this to a separate helper function? Same goes for the other ones below.

Then please apply the above feedback on error messages and variables names as well and I think we're done.